keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/tensor/TensorFitnessFunctionWithGradient.java

```diff
@@ -52,7 +77,11 @@ public MultivariateFunction fitness() {
         for (Vertex<DoubleTensor> vertex : latentVertices) {
             DoubleTensor tensor = diffs.get(vertex.getId());
             if (tensor != null) {
-                tensors.add(tensor);
+                if (TensorShape.isScalar(vertex.getShape()) && !TensorShape.isScalar(tensor.getShape())) {
```

Comment:
This is quite nested and could potentially be made more expressive if this were extracted into a separate function.

Reply:
Yup, agreed.